CounterpointGenerator/CounterpointGenerator/Generator.cs

```diff
@@ -40,21 +40,36 @@ private List<MelodyLine> GenerateCounterpointForNoteStack(MelodyLine m, Note? pr
 
         }
 
-        private List<Note> CounterpointForNote(Note n, Note prevN, Note preCtp)
+        private List<Note> CounterpointForNote(Note n, Note prevN, Note preCtp, int melodyLength, int generateCount)
         {
             List<Note> possibleNotes = GenerateNoteAtRegion(n);
-            List<Note> possibleNotesAfterRules = new List<Note>();
-            /*foreach(IRules r in Dictionary[input.userPreference])
+            // Possibilities, currentNote, position, length, previousCantus, previousCounterpoint
+            RuleInput ruleInput = new RuleInput(possibleNotes, n, generateCount, melodyLength, prevN, preCtp);
+
+            ruleApplier = new RuleApplier();
+            if (ruleApplier.GenerateSet())
+            {
+                ruleApplier.Applicator(ruleInput);
+            }
+            else
             {
-                List<Note> newPossibleNotes = r.Apply(possibleNotes);
-                possibleNotesAfterRules.AddRange(newPossibleNotes);
-            }*/
-            return possibleNotesAfterRules;
+                throw new Exception("Ruleset did not generate properly!");
+            }
+
+            return ruleInput.Possibilities;
         }
 
         private List<Note> GenerateNoteAtRegion(Note n)
         {
-            throw new NotImplementedException();
+            int upperPitch = n.Pitch + 13; // +1 octave
```

Comment:
Avoid "magic numbers".
https://www.c-sharpcorner.com/UploadFile/b7dc95/writing-clean-code-and-magic-numbers/

Reply:
Since 13 (octave) is project-wide consider doing something like in the top answer here. https://stackoverflow.com/questions/2032978/how-to-manage-project-wide-constants-in-a-project-using-c

Reply:
Resolved a9a7fdc